### PR TITLE
Py stacktrace

### PR DESCRIPTION
**Pull request by tpoterba**
This PR breaks "fatal" out into two functions, "fatal" and "abort".  Fatal is for unexpected error handling, and produces a python stacktrace, but 'abort' is for handled, expected errors (like invalid method inputs) and does not generate a stack trace in python.

### CODE REVIEW

**Comment by tpoterba** [comment]
This PR breaks "fatal" out into two functions, "fatal" and "abort".  Fatal is for unexpected error handling, and produces a python stacktrace, but 'abort' is for handled, expected errors (like invalid method inputs) and does not generate a stack trace in python.

**Comment by cseed** [comment]
I haven't read over this, but I don't like the behavior.  Assert and friends are for unexpected errors, and fatal is for expected errors.  How is abort different from assert?

All errors should give full JVM + Python stack traces.  I see this necessary for two reasons: It makes it much easier for users to report bugs to us, which means they get faster turnaround and we spend less time going back and forth about log files (which usually were ephemeral or they've overwritten) and often "expected" bugs are actually correct behavior on the user's end and a bug on our side, but no context is given for us to diagnose the real problem.

For usability, it is obviously best if the user-visible error appears at the bottom.

**Comment by tpoterba** [comment]
I don't see the utility in creating an unnecessary stack trace to see 'method "variant QC" requires a split dataset'

I think there is value in having clear, short, stack-trace-free error messages when it's clear what the problem is and what the user needs to do. I think that printing unnecessary stack traces will cause users to view hail even more as a tool in development, and will be more inclined to ask us about errors rather than try to figure out how whether they made a simple mistake using the interface.

**Comment by cseed** [comment]
We decided not to do this.

**Comment by tpoterba** [comment]
Reopening, every error will produce a stack trace.

**Comment by tpoterba** [comment]
Got rid of @handle_py4j, as well, which removes our dependence on decorator 👍

**Comment by tpoterba** [comment]
here are some example error msgs:
```
In [2]: vds.linreg([])
---------------------------------------------------------------------------
FatalError                                Traceback (most recent call last)
<ipython-input-2-b8bbc41a5ebd> in <module>()
----> 1 vds.linreg([])

/Users/tpoterba/hail/python/hail/dataset.py in linreg(self, y, covariates, root, min_ac, min_af)
   2216         """
   2217
-> 2218         jvds = self._jvdf.linreg(y, jarray(env.jvm.java.lang.String, covariates), root, min_ac, min_af)
   2219         return VariantDataset(self.hc, jvds)
   2220

/Users/tpoterba/spark-2.0.2-bin-hadoop2.7/python/lib/py4j-0.10.3-src.zip/py4j/java_gateway.py in __call__(self, *args)
   1131         answer = self.gateway_client.send_command(command)
   1132         return_value = get_return_value(
-> 1133             answer, self.gateway_client, self.target_id, self.name)
   1134
   1135         for temp_arg in temp_args:

/Users/tpoterba/spark-2.0.2-bin-hadoop2.7/python/pyspark/sql/utils.pyc in deco(*a, **kw)
     61     def deco(*a, **kw):
     62         try:
---> 63             return f(*a, **kw)
     64         except py4j.protocol.Py4JJavaError as e:
     65             s = e.java_exception.toString()

/Users/tpoterba/hail/python/hail/java.py in deco(*a, **kw)
    113             if e.args[0].startswith('An error occurred while calling'):
    114                 msg = 'An error occurred while calling into JVM, probably due to invalid parameter types'
--> 115                 raise FatalError('%s\n\nJava stack trace:\n%s\n\nERROR SUMMARY: %s' % (msg, e.message, msg))
    116
    117     return deco

FatalError: An error occurred while calling into JVM, probably due to invalid parameter types

Java stack trace:
An error occurred while calling o29.linreg. Trace:
py4j.Py4JException: Method linreg([class java.util.ArrayList, class [Ljava.lang.String;, class java.lang.String, class java.lang.Integer, class java.lang.Double]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:318)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:326)
	at py4j.Gateway.invoke(Gateway.java:272)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)



ERROR SUMMARY: An error occurred while calling into JVM, probably due to invalid parameter types```

**Comment by tpoterba** [comment]
```
In [4]: hc.import_vcf('src/test/resources/malformed.vcf').count()
hail: info: Coerced sorted dataset
---------------------------------------------------------------------------
FatalError                                Traceback (most recent call last)
<ipython-input-4-955ea6a16c80> in <module>()
----> 1 hc.import_vcf('src/test/resources/malformed.vcf').count()

/Users/tpoterba/hail/python/hail/dataset.py in count(self, genotypes)
   1127         """
   1128
-> 1129         return dict(self._jvdf.count(genotypes).toJavaMap())
   1130
   1131     def deduplicate(self):

/Users/tpoterba/spark-2.0.2-bin-hadoop2.7/python/lib/py4j-0.10.3-src.zip/py4j/java_gateway.py in __call__(self, *args)
   1131         answer = self.gateway_client.send_command(command)
   1132         return_value = get_return_value(
-> 1133             answer, self.gateway_client, self.target_id, self.name)
   1134
   1135         for temp_arg in temp_args:

/Users/tpoterba/spark-2.0.2-bin-hadoop2.7/python/pyspark/sql/utils.pyc in deco(*a, **kw)
     61     def deco(*a, **kw):
     62         try:
---> 63             return f(*a, **kw)
     64         except py4j.protocol.Py4JJavaError as e:
     65             s = e.java_exception.toString()

/Users/tpoterba/hail/python/hail/java.py in deco(*a, **kw)
    109             # deepest = env.jutils.deepestMessage(e.java_exception)
    110             # msg = env.jutils.getMinimalMessage(e.java_exception)
--> 111             raise FatalError('%s\n\nJava stack trace:\n%s\n\nERROR SUMMARY: %s' % (deepest, full, deepest))
    112         except py4j.protocol.Py4JError as e:
    113             if e.args[0].startswith('An error occurred while calling'):

FatalError: HailException: malformed.vcf: caught htsjdk.tribble.TribbleException$InternalCodecException: The allele with index 10026357 is not defined in the REF/ALT columns in the record
  offending line: 20	10026348	.	A	G	7535.22	PASS	HWP=1.0;AC=2;culprit=Inbreedi...

Java stack trace:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 3.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3.0 (TID 3, localhost): is.hail.utils.HailException: malformed.vcf: caught htsjdk.tribble.TribbleException$InternalCodecException: The allele with index 10026357 is not defined in the REF/ALT columns in the record
  offending line: 20	10026348	.	A	G	7535.22	PASS	HWP=1.0;AC=2;culprit=Inbreedi...
	at is.hail.utils.ErrorHandling$class.fatal(ErrorHandling.scala:10)
	at is.hail.utils.package$.fatal(package.scala:20)
	at is.hail.utils.TextContext.wrapException(Context.scala:15)
	at is.hail.utils.WithContext.map(Context.scala:27)
	at is.hail.io.vcf.LoadVCF$$anonfun$14$$anonfun$apply$7.apply(LoadVCF.scala:301)
	at is.hail.io.vcf.LoadVCF$$anonfun$14$$anonfun$apply$7.apply(LoadVCF.scala:301)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.sparkextras.OrderedRDD$$anonfun$apply$7$$anon$2.hasNext(OrderedRDD.scala:210)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.util.Utils$.getIteratorSize(Utils.scala:1763)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1134)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1134)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1899)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1899)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:86)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1454)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1442)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1441)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1441)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:811)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:811)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:811)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1667)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1622)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1611)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:632)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1873)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1886)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1899)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1913)
	at org.apache.spark.rdd.RDD.count(RDD.scala:1134)
	at is.hail.variant.VariantSampleMatrix.countVariants(VariantSampleMatrix.scala:810)
	at is.hail.variant.VariantDatasetFunctions$.count$extension(VariantDataset.scala:504)
	at is.hail.variant.VariantDatasetFunctions.count(VariantDataset.scala:494)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:237)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)is.hail.utils.HailException: malformed.vcf: caught htsjdk.tribble.TribbleException$InternalCodecException: The allele with index 10026357 is not defined in the REF/ALT columns in the record
  offending line: 20	10026348	.	A	G	7535.22	PASS	HWP=1.0;AC=2;culprit=Inbreedi...
	at is.hail.utils.ErrorHandling$class.fatal(ErrorHandling.scala:10)
	at is.hail.utils.package$.fatal(package.scala:20)
	at is.hail.utils.TextContext.wrapException(Context.scala:15)
	at is.hail.utils.WithContext.map(Context.scala:27)
	at is.hail.io.vcf.LoadVCF$$anonfun$14$$anonfun$apply$7.apply(LoadVCF.scala:301)
	at is.hail.io.vcf.LoadVCF$$anonfun$14$$anonfun$apply$7.apply(LoadVCF.scala:301)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.sparkextras.OrderedRDD$$anonfun$apply$7$$anon$2.hasNext(OrderedRDD.scala:210)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.util.Utils$.getIteratorSize(Utils.scala:1763)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1134)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1134)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1899)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1899)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:86)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)



ERROR SUMMARY: HailException: malformed.vcf: caught htsjdk.tribble.TribbleException$InternalCodecException: The allele with index 10026357 is not defined in the REF/ALT columns in the record
  offending line: 20	10026348	.	A	G	7535.22	PASS	HWP=1.0;AC=2;culprit=Inbreedi...
```

**Comment by tpoterba** [comment]
Okay, this isn't passing tests because Py4J actually catches Py4JErrors and uses them to stop iteration on java collections. Very frustrating.

File: src/main/scala/is/hail/HailContext.scala [hunk 1]
```diff
@@ -249,12 +249,6 @@ class HailContext private(val sc: SparkContext,
     if (inputs.isEmpty)
       fatal("arguments refer to no files")
 
-    inputs.foreach { input =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/HailContext.scala]: could be in a different PR, I guess. This prevents loading .gen.gz files, which are common.

File: src/main/scala/is/hail/utils/Context.scala [hunk 2]
```diff
@@ -6,13 +6,16 @@ abstract class Context {
...
   def wrapException(e: Throwable): Nothing = {
-    val msg = e match {
-      case _: FatalException => e.getMessage
-      case _ => s"caught $e"
+    e match {
+      case _: HailException =>
+        fatal(
+          s"""$file${ position.map(ln => ":" + (ln + 1)).getOrElse("") }: ${ e.getMessage }
+             |  offending line: @1""".stripMargin, line)
+      case _ => e.getMessage
```
Comment (danking) [inline src/main/scala/is/hail/utils/Context.scala]: this has no effect, right?

File: src/main/scala/is/hail/utils/Context.scala [hunk 3]
```diff
@@ -6,13 +6,16 @@ abstract class Context {
...
+        fatal(
+          s"""$file${ position.map(ln => ":" + (ln + 1)).getOrElse("") }: caught ${ e.getClass.getName() }: ${ e.getMessage }
+             |  offending line: @1""".stripMargin, line)
```
Comment (danking) [inline src/main/scala/is/hail/utils/Context.scala]: I know your PR doesn't change this, but it'd be nice if we preserved the stack traces from `e` but making them the `cause` of the `FatalException` (`cause` is inherited from `RuntimeException`).

Comment (tpoterba) [inline src/main/scala/is/hail/utils/Context.scala]: Yeah, it's easy to set the cause, but it requires a bit more work because we want the fatal message to be in the "Error summary" block. Will save for a future PR

File: src/main/scala/is/hail/utils/Context.scala (hunk 2 above)
Comment (tpoterba) [inline src/main/scala/is/hail/utils/Context.scala]: oops, yup

**Comment by tpoterba** [comment]
addressed comments
